File: python/euler3.py
import math

def sieve(n):
    primes = []
    A = [True for x in range(n+1)]
    A[0] = A[1] = False # 0 and 1 are not prime
    for i in range(2, math.ceil(math.sqrt(n))):
        if A[i]:
            for j in range(i**2, n+1, i):
                A[j] = False   
    for i in range(len(A)):
        if A[i]:
            primes.append(i)
    return primes

def factor(n):
    primes = sieve(n)
    factors = []
    while n != 1:
        for i in range(len(primes)):
            if n % primes[i] == 0:
                n /= primes[i]
                factors.append(primes[i])
                break
    return factors

# Shallow test
print(sieve(10) == [2, 3, 5, 7])    
print(factor(200) == [2,2,2,5,5])
print(factor(13195) == [5,7,13,29])

print(sieve(600851475143))

File: javascript/euler3.js
function factors(n, MAX_SIEVE_SIZE = 2 ** 16) {
    let factors = [];
    let primes = [];
    let sieve = [];

    for (let i = 0; i < MAX_SIEVE_SIZE; i++) {
        sieve[i] = true;
    }

    sieve[0] = sieve[1] = false;

    let p = 2;

    do {
        primes.push(p);

        if (n % p == 0) {
            factors.push(p);
        }

        delete sieve[p];

        for (let i = 2*p; i < MAX_SIEVE_SIZE; i += p) {
            sieve[i] = false;
        }

        do {
            p++;
        } while (sieve[p] == false);
    } while (p < MAX_SIEVE_SIZE);

    return factors;
}

(() => {
    console.time('prime');
    console.log(factors(600851475143));
    console.timeEnd('prime');
})();
